packages/vite/src/node/build.ts

```diff
@@ -360,6 +360,37 @@ export function resolveBuildOptions(
         : defaultModulePreload
   }
 
+  // Auto-resolve dependencies if not explicitly specified
+  if (raw && raw.lib !== false && raw.rollupOptions?.external == null) {
+    const formats = raw.lib?.formats ?? []
+    // If umd or iife is specified, we need to bundle all dependencies and cannot use external automatically
+    const isUMD = formats.includes('umd')
+    const isIIFE = formats.includes('iife')
```

Comment:
Default value of `lib.formats` is `['es', 'cjs']` or `['es', 'umd']`.
https://github.com/vitejs/vite/blob/75dbca2789b0aca0729f74b0e69ed5a9c08eb1df/packages/vite/src/node/build.ts#L795-L797

The auto-externalization will work if `lib.formats` are not set, even if the actual format included `umd`.

Also `configResolved` hook could change `lib.formats` after this part.
https://github.com/vitejs/vite/blob/75dbca2789b0aca0729f74b0e69ed5a9c08eb1df/packages/vite/src/node/config.ts#L529
https://github.com/vitejs/vite/blob/75dbca2789b0aca0729f74b0e69ed5a9c08eb1df/packages/vite/src/node/config.ts#L716-L724

In addition, we might need to think about what happens if `build.rollupOptions` is an array. For example,
```js
[
  { dir: './foo', format: 'es' },
  { dir: './foo', format: 'umd' }
}
```

Reply:
> In addition, we might need to think about what happens if `build.rollupOptions` is an array.

`resolved.rollupOptions.external = allDependencies` didn't result in an error :thinking: 
Is the typedef wrong?

Reply:
I wanted to write `build.rollupOptions.output`, not `build.rollupOptions`. 🤦‍♂️

Reply:
@sapphi-red I'm not that sure about your points and I also don't have that deep knowledge in Vite code base
The PR currently solves the main issue I pointed out in https://github.com/vitejs/vite/discussions/10582#discussion-4493470

- if in any kind of way umd or iife is configured, the auto detection is turned of
- if the config is more complex, the auto detection is turned of
- if no `package.json` could get detected, the auto detection is turned of
- if no relevant dependencies were found, the auto detection is turned of

So only in the special case were you didn't configured `rollupOptions.external` but `lib`, the detection will take effect.
You are always have the possibility to manually configure `rollupOptions.external` and nothing of this detection would even start.

So maybe we can merge and iterate later if someone has more specific request to this feature. If there is no bug I currently overlook in any way, I feel really safe to merge this as is.
And even if, configure `rollupOptions.external` manually and you are as you need to do right now in Vite v3.

Reply:
> if in any kind of way umd or iife is configured, the auto detection is turned of

I mean this is not working in some cases.
I guess auto detection shouldn't take effect for these configs.
But auto detection does take effect with this PR's implementation.

1. ```js
   // vite.config.js
   export default {
     build: {
       lib: {
         entry: './src/main.js',
         name: 'main'
         // format defaults to `['es', 'umd']`
         // because it doesn't have multiple entries
       }
     }
   }
   ```
1. ```js
   // vite.config.js
   export default {
     build: {
       lib: {
         entry: './src/main.js',
         name: 'main',
         formats: ['es']
       }
     },
     plugins: [
       // a plugin that adds umd format in configResolved hook
       {
         name: 'foo',
         configResolved(config) {
           config.build.lib.formats.push('umd')
         }
       }
     ]
   }
   ```
1. ```js
   {
     build: {
       lib: {
         entry: path.resolve(__dirname, 'src/main.js'),
         name: 'MyLib'
       },
       rollupOptions: {
         output: [
           // format is set in rollupOptions.output instead of lib.formats
           { format: 'es' },
           { name: 'MyLib', format: 'umd' }
         ]
       }
     }
   }
   ```

Reply:
Ah, now I understand 👌
Thanks for the examples, they make it much more clear to me

Reply:
@sapphi-red If I would check explicitly for `raw.lib.formats` to exist and only then do the auto-detection, would this be enough for now?

I mean like this:
```ts
  // Auto-resolve dependencies if not explicitly specified
  if (
    raw &&
    !!raw.lib &&
    raw.lib.formats &&
    raw.lib.formats.length > 0 &&
    raw.rollupOptions?.external == null
  ) {
    const formats = raw.lib.formats
    // If umd or iife is specified, we need to bundle all dependencies and cannot use external automatically
    const isUMD = formats.includes('umd')
    const isIIFE = formats.includes('iife')
    if (!isUMD && !isIIFE) {
```

---

But I'm absolutely not sure how to handle `configResolved`
It goes so deep into Vite and I just don't have the knowledge
So if this is the blocking part here and I cannot get further help or even peer programming, I'm lost
Maybe I just write a simple plugin that uses this `configResolved` instead and do this auto-detection stuff via that plugin :thinking: 
In that case, I can then also be sure, that I only use the plugin if I know what I want, but do not forget to add a dependency when I add a new one to my project

Reply:
Maybe moving the code after this part?
https://github.com/vitejs/vite/blob/75dbca2789b0aca0729f74b0e69ed5a9c08eb1df/packages/vite/src/node/build.ts#L600-L614
I guess it will work for all of the cases I said.

Reply:
suggestions applied
tested locally with a playground and all 3 + my own configuration, looks like it works